### ISSUE TITLE
tf.keras.layers.Embedding updating embeddings of nearby indices even though they were not updated in the code

### ISSUE THREAD

**Issue by piyush-wynk** [tensorflow/tensorflow#47187]
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Ubuntu 16.04
- TensorFlow version (use command below): 2.4.1
- Python version: 3.6.9

I am using tf.keras.layers.embedding on a dictionary size of 20 million. For indices greater than 16777216, updates happening in one embedding gets reflected in some other embedding .i.e I am getting same final embeddings for indices above 16777216 in groups of 3.
eg. final embeddings for 16777216, 16777217, 16777218 are equal, 16777220, 16777221, 16777222 are equal and so on.

Somehow higher indices are getting mapped to same entry in the look up table, or may be some hash collision.

Simplified reproducible code can be found [here.](https://colab.research.google.com/drive/1m-zz4mlPMqOZ_fpz9GpwHTllqH3uqM4g?usp=sharing)

<img width="1422" alt="Screenshot 2021-02-16 at 8 36 26 PM" src="https://user-images.githubusercontent.com/69950926/108081744-4d0cf700-7097-11eb-847c-b93e7897e089.png">

Also, on a side note, this behaviour is not observed in tf version 2.1.0, embedding layer works fine there.
Any help will be really appreciated, as this is resulting in serious performance issue.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3, [TF v2.4](https://colab.research.google.com/gist/amahendrakar/3d5e8681055b8b4ba7c85e2421907b1f/47187.ipynb) and TF-nightly. 

And as mentioned, code works fine with [TF v2.1](https://colab.research.google.com/gist/amahendrakar/6bfb14e802fd9f35a6c8ef497a920a5b/47187-2-1.ipynb). Please check the linked gist for reference. Thanks!

**Comment by fchollet** [comment]
Chiming in here. This isn't exactly a bug. Though we may be able to do something on the Keras side to avoid users running into this issue.

As you may be aware, the default dtype for all Keras model inputs is `float32`. This is well documented.

In your model you don't specify an input dtype. Therefore you default to `float32`.

`float32` doesn't have the numerical precision to hold large integer values. Hence we get numerical collisions.

When using an Embedding layer with large input values, make sure to specify the input dtype as `int64`.

Here's an example that works:

```python
user_model = tf.keras.Sequential([
    tf.keras.Input((1,), dtype='int64'),
    tf.keras.layers.Embedding(20000000, 10)
])

user_model.compile(optimizer=tf.keras.optimizers.Adagrad(learning_rate=0.2),loss='mse')
```

Action item for the Keras team: consider refusing to process `float32` inputs if the embedding space is too large.

**Comment by piyush-wynk** [comment]
> Chiming in here. This isn't exactly a bug. Though we may be able to do something on the Keras side to avoid users running into this issue.
> 
> As you may be aware, the default dtype for all Keras model inputs is `float32`. This is well documented.
> 
> In your model you don't specify an input dtype. Therefore you default to `float32`.
> 
> `float32` doesn't have the numerical precision to hold large integer values. Hence we get numerical collisions.
> 
> When using an Embedding layer with large input values, make sure to specify the input dtype as `int64`.
> 
> Here's an example that works:
> 
> ```python
> user_model = tf.keras.Sequential([
>     tf.keras.Input((1,), dtype='int64'),
>     tf.keras.layers.Embedding(20000000, 10)
> ])
> 
> user_model.compile(optimizer=tf.keras.optimizers.Adagrad(learning_rate=0.2),loss='mse')
> ```
> 
> Action item for the Keras team: consider refusing to process `float32` inputs if the embedding space is too large.



Thank you for the input. After specifying dtype, the problem of initialisation by same vector seems to be gone. 

But, there still seems to be some problem with higher indices. I am using embedding vocabulary of around 40 million data points, and for test data with indices greater than 30 million, getting very poor metric. But when I reverse the indices during training, such that the test set has same data (with reversed indices eg. data point with index 40 million will now have index 0, and index 0 will now be 40 million) and training data in a batch also remains same, but the indices of test instances are now less than 10 million, great improvement in metric is observed i.e the model is performing good for smaller indices, but poor for larger indices, irrespective of the data.

Can not figure out why this is happening .

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF 2.6.0-dev20210528 ,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/0377f18f18e1ba42232a0c9468115455/untitled34.ipynb#scrollTo=fLPQgbDHXGKo)...Thanks !

**Comment by sachinprasadhs** [comment]
I still see the issue existing in updating embeddings of nearby indices in Tensorflow 2.7. Since development of keras moved to separate repository https://github.com/keras-team/keras/issues

Please post this issue on keras-team/keras repo.
To know more see;
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
Thank you!
